### PR TITLE
Try using axe disableFrame function to resolve Axe end to end test error

### PR DESCRIPTION
**Pull request by talldan**
## Description
Currently axe core tests are failing intermittently on this test:
```
Template Part › Template part placeholder › Should insert template part when preview is selected

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Failed to inject axe-core into frame (about:blank)"],["Failed to inject axe-core into frame (about:blank)"]
```

There's an error injecting axe core scripts into iframes created by the block preview.

https://github.com/WordPress/gutenberg/pull/26527 tried to use the exclude option to make axe ignore block previews, but Axe still tries to inject the script.

Axe has a `disableFrame` function that looks like it prevents axe from inject scripts into specified iframes. Its undocumented (https://github.com/dequelabs/axe-core-npm/pull/39), so not sure how we'd feel about using it.

This PR tests it out.

### CODE REVIEW

**Comment by gziolo** [comment]
> There's an error injecting axe core scripts into iframes created by the block preview.

It would explain why it fails also for the demo test:
https://github.com/WordPress/gutenberg/blob/467175e6a17d412a77b53a9df36023e319ed5515/packages/e2e-tests/specs/local/demo.test.js#L42
It triggers the preview as well. The way to go would be rather to find a way to ensure that the page that is about to be validated contains the block editor. We surely don't want to validate `about:blank` page.

**Comment by talldan** [comment]
> It triggers the preview as well.

@gziolo I didn't realise that. I thought it just loaded the demo and then checked the dirty state.

File: packages/e2e-tests/config/setup-test-framework.js [hunk 1]
```diff
@@ -206,6 +206,10 @@ async function runAxeTestsForBlockEditor() {
 			'aria-required-parent',
 			'frame-title',
 		],
+		disabledFrames: [
+			// Ignore block previews.
+			'[aria-hidden="true"]',
```
Comment (talldan) [inline packages/e2e-tests/config/setup-test-framework.js:211]: The iframe in question is added by `react-resize-aware` as part of the `useResizeObserver` hook and has `aria-hidden="true"` on it, but as a general rule it seems like axe shouldn't be trying to run tests on the contents of iframes that are aria-hidden.

**Comment by gziolo** [comment]
> > It triggers the preview as well.
> 
> @gziolo I didn't realise that. I thought it just loaded the demo and then checked the dirty state.

Ah, wait. I misinterpreted this code. Anyway, something goes wrong there as well. So maybe there is something strange happening that creates iframes that are empty? It might be also an issue with axe itself, since we upgraded to v4.

**Comment by talldan** [comment]
The stack trace does seem to indicate it's caused by axe. The embed mocking in that test creates an empty iframe, so that could be related?

The other thought I had is that it doesn't like the multiple `visitAdminPage` calls are causing an unexpected navigation, but then `visitAdminPage` works pretty well elsewhere.

**Comment by talldan** [comment]
Well, all checks have passed twice in a row now, which seems like more success than we've had lately?

I might just want to double-check that I haven't disabled axe tests entirely 😅

**Comment by gziolo** [comment]
> Well, all checks have passed twice in a row now, which seems like more success than we've had lately?
> 
> I might just want to double-check that I haven't disabled axe tests entirely 😅

Let's make sure that those axe tests still verify enabled rules.

How about this new option is marked as experimental for the start so we could validate if it solves the issue? This way we won't need to document it inside `@wordpress/jest-puppeteer-axe` until we are sure this API is solid. By the way, this package is used in the Storybook addon (https://www.npmjs.com/package/@storybook/addon-storyshots-puppeteer).

**Comment by gziolo** [comment]
I commented out some skipped axe rules to see if axe still works. Feel free to remove 01eff11b02b74376384f6de3ee1ba8bc3b57a94a once we know :)

**Comment by gziolo** [comment]
There are valid failures reported which means that everything works perfectly fine with the changes proposed. I'm happy to approve this PR when `disabledFrames` is updated to `__experimentalDisabledFrames` and my test commit is removed.

File: packages/e2e-tests/config/setup-test-framework.js [hunk 2]
```diff
@@ -193,19 +193,23 @@ async function runAxeTestsForBlockEditor() {
...
 			'duplicate-id',
-			'label',
-			'landmark-one-main',
+			// 'label',
+			// 'landmark-one-main',
 			'link-name',
-			'listitem',
-			'region',
+			// 'listitem',
+			// 'region',
```
Comment (Addison-Stavlo) [inline packages/e2e-tests/config/setup-test-framework.js:204]: I might have missed it in the discussion.  Was there a reason we are commenting out a handful of the disabled rules?

**Review by Addison-Stavlo — COMMENTED** [review]
This proposed solution with the `disabledFrames` looks good to me!  Also supporting the `__experimental` recommendation.

**Review by Addison-Stavlo — COMMENTED** [review]
Also worth noting #26527 seems to be passing now (3x in a row at least) that I updated the selector for block previews from the container to the content.  So that might still make sense as an alternative (or addition) to this?

Its a bit less 'experimental' in nature than what we are doing here with the undocumented `disableFrame`, but I agree the approach here also makes sense and that disabling everything under `[aria-hidden="true"]` seems like it should be done in general for axe tests. 🤔

**Comment by talldan** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/26527. We can always come back to this if the problem reoccurs.

File: packages/e2e-tests/config/setup-test-framework.js (hunk 1 above)
Comment (talldan) [reply packages/e2e-tests/config/setup-test-framework.js:214]: Added a comment to explain this.

**Comment by talldan** [comment]
> There are valid failures reported which means that everything works perfectly fine with the changes proposed. I'm happy to approve this PR when disabledFrames is updated to __experimentalDisabledFrames and my test commit is removed.

Agree it should be experimental for now. I've made that change.

**Comment by talldan** [comment]
Ah, well on the last run this also just failed with the same iframe error:
```
FAIL specs/experiments/template-part.test.js (21.914s)
  ● Template Part › Template part placeholder › Should insert template part when preview is selected

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Failed to inject axe-core into frame (about:blank)"],["Failed to inject axe-core into frame (about:blank)"]
```

So I guess this won't solve the issue either.

**Comment by talldan** [comment]
Another alternative, we can ignore the logged message - https://github.com/WordPress/gutenberg/pull/26590

**Pull request closed without merging**